kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs

```diff
@@ -0,0 +1,301 @@
+//   * The values and cell contents are in little-endian order, which
+//     is important for compatibility with other gates.
+//   * The witness sections for the multi range check gates should be set up
+//     so that the last range checked value is the MS limb of the respective
+//     foreign field element. For example, given foreign field element q
+//     such that
+//
+//         q = q0 + 2^88 * q1 + 2^176 * q2
+//
+//     and multi-range-check gate witness W, where W[r][c] accesses row r
+//     and column c, we should map q to W like this
+//
+//         W[0][0] = q0
+//         W[1][0] = q1
+//         W[2][0] = q2
+//
+//     so that most significant limb, q2, is in W[2][0].
+//
+fn create_layout<F: PrimeField>() -> [[Box<dyn WitnessCell<F>>; COLUMNS]; 2] {
+    [
+        // ForeignFieldMul row
+        [
+            // Copied for multi-range-check
+            VariableCell::create("left_input0"),
+            VariableCell::create("left_input1"),
+            VariableCell::create("left_input2"),
+            // Copied for multi-range-check
+            VariableCell::create("right_input0"),
+            VariableCell::create("right_input1"),
+            VariableCell::create("right_input2"),
+            VariableCell::create("carry1_lo"), // Copied for multi-range-check
+            VariableCell::create("carry1_hi"), // 12-bit lookup
+            VariableCell::create("scaled_carry1_hi"), // 12-bit lookup
+            VariableCell::create("carry0"),
+            VariableCell::create("quotient0"),
+            VariableCell::create("quotient1"),
+            VariableCell::create("quotient2"),
+            VariableCell::create("quotient_bound_carry01"),
+            VariableCell::create("quotient_bound_carry2"),
+        ],
+        // Zero row
+        [
+            // Copied for multi-range-check
+            VariableCell::create("remainder0"),
+            VariableCell::create("remainder1"),
+            VariableCell::create("remainder2"),
+            VariableCell::create("quotient_bound01"),
+            VariableCell::create("quotient_bound2"),
+            VariableCell::create("product1_lo"), // Copied for multi-range-check
+            VariableCell::create("product1_hi_0"), // Copied for multi-range-check
+            VariableCell::create("product1_hi_1"),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+            ConstantCell::create(F::zero()),
+        ],
+    ]
+}
+
+pub fn compute_bound(x: &BigUint, neg_foreign_field_modulus: &BigUint) -> BigUint {
+    let x_bound = x + neg_foreign_field_modulus;
+    assert!(x_bound < BigUint::binary_modulus());
+    x_bound
+}
+
+fn compute_witness_variables<F: PrimeField>(
+    products: &[BigUint; 3],
+    remainder: &[BigUint; 3],
+) -> [F; 7] {
```

Comment:
I am inclined to think that we may want to have structs for this type of return types where position really has a different meaning and making mistakes could be common otherwise. Same applies to FFAdd, by the way

Reply:
That would be a concern if this function was used in more than one place, but this is only used once, is not a public interface and we'd have a hard time passing any tests if any outputs are misaligned.  It just feels overkill to create a struct if it's only ever used once.